docs/source/configuration.rst
.. _configuration-file:

Configuration file
==================

The *babelizer* requires a single *toml*-formatted configuration file
that describes the library to wrap.
This file is typically named ``babel.toml``.
An example of a ``babel.toml`` file:

.. code:: toml

  # See https://babelizer.readthedocs.io/ for more information

  # Describe the library being wrapped.
  [library.Monorail]
  language = "c"
  library = "bmimonorail"
  header = "monorail.h"
  entry_point = "register_monorail"

  # Describe compiler options to build the library being wrapped.
  [build]
  undef_macros = []
  define_macros = []
  libraries = []
  library_dirs = []
  include_dirs = []
  extra_compile_args = []

  # Describe the newly wrapped package.
  [package]
  name = "springfield_monorail"
  requirements = ["three_million_dollars"]

  # Provide author and package information.
  [info]
  github_username = "lyle-lanley"
  package_author = "Lyle Lanley"
  package_author_email = "lyle@monorail.com"
  package_license = "MIT License"
  summary = '''
  Well, sir, there's nothing on Earth like a genuine,
  bona fide, electrified, six-car monorail. What'd I say?
  Monorail! What's it called? Monorail! That's right! Monorail!
  '''

  # Set continuous integration options.
  [ci]
  python_version = [
      "3.10",
      "3.11",
      "3.12",
  ]
  os = [
      "linux",
      "mac",
      "windows",
  ]

You can generate *babelizer* configuration files using the ``babelize sample-config`` command.
For example, the above ``babel.toml`` was generated with:

.. code:: bash

  babelize sample-config > babel.toml

Library section
---------------

The *library* section provides information about the library being *babelized*.

Name
""""

The name of the *babelized* class is integrated into the *library* table header;
in the example above, ``Monorail``.
This will be a Python class,
so it should follow Python naming conventions such as camel-case typing.

Language
""""""""

The programming language of the library (possible values are "c", "c++",
"fortran", and "python").

Library
"""""""

The name of the BMI library to wrap.
This is the text passed to the linker through the `-l` option;
for example, use "bmimonorail" for a library ``libbmimonorail.a``.

Header
""""""

The name of the header file (``.h``, ``.hxx``) declaring the BMI class.
This option is only needed when wrapping C and C++ libraries.

Entry point
"""""""""""

The name of the BMI entry point into the library.
For object-oriented languages,
this is typically the name of a class that implements the BMI.
For procedural languages,
this is typically a function.

An example of a C++ library (*bmi_child*), exposing a class *BmiChild* (that
implements a BMI) might look like the following:

.. code:: toml

  [library.Child]
  language = "c++"
  library = "bmi_child"
  header = "bmi_child.hxx"
  entry_point = "BmiChild"

whereas a C library (*bmi_cem*), exposing a function *register_bmi_cem* (that
implements a BMI) might look like:

.. code:: toml

   [library.Cem]
   language = "c"
   library = "bmi_cem"
   header = "bmi_cem.h"
   entry_point = "register_bmi_cem"

Build section
-------------

In the *build* section, specify flags to pass to the compiler
when building the extension.

Package section
---------------

The *package* section provides the name and extra requirements needed to build the *babelized* library.

Name
""""

The name to use for the Python package output from the *babelizer*.

Requirements
""""""""""""

A list of packages required by the library being wrapped. For example, the
following indicates that the packages *foo* and *bar* are dependencies
for the package.

.. code:: toml

  [package]
  requirements = [ "foo", "bar",]

Info section
------------

Descriptive information about the package.

Github username
"""""""""""""""

The GitHub username or organization where this package will be hosted. This
is used in generating links to the CI, docs, etc.

Author
""""""

Author of the wrapped package. Note that this is not the author of the
library being wrapped, just the code generated by the *babelizer*.

Email
"""""

Contact email to use for the wrapped package.

License
"""""""

Specify the Open Source license for the wrapped package. Note that this is not the
license for the library being wrapped, just for the code generated by the *babelizer*.

Summary
"""""""

A short description of the wrapped library.

CI section
----------

Information about how to set up continuous integration.

Python version
""""""""""""""

A list of Python versions to build and test the generated project with.

Operating system
""""""""""""""""

A list of operating systems on which to build the generated project. Supported values are
*linux*, *mac*, and *windows*.

Example configuration file
--------------------------

Below is an example of a *babelizer* configuration file that describes a shared library,
written in C. In this example, the library, *bmi_hydrotrend*, exposes the
function *register_bmi_hydrotrend* that implements a BMI for a component
called *hydrotrend*.

.. code:: toml

    [library.Hydrotrend]
    language = "c"
    library = "bmi_hydrotrend"
    header = "bmi_hydrotrend.h"
    entry_point = "register_bmi_hydrotrend"

    [build]
    undef_macros = []
    define_macros = []
    libraries = []
    library_dirs = []
    include_dirs = []
    extra_compile_args = []

    [package]
    name = "pymt_hydrotrend"
    requirements = ["hydrotrend"]

    [info]
    github_username = "pymt-lab"
    package_author = "csdms"
    package_author_email = "csdms@colorado.edu"
    package_license = "MIT License"
    summary = "PyMT plugin for hydrotrend"

    [ci]
    python_version = ["3.10", "3.11", "3.12"]
    os = ["linux", "mac", "windows"]

For other examples of *babelizer* configuration files,
see the directories under the `external/tests <https://github.com/csdms/babelizer/tree/develop/external/tests>`_
directory of the *babelizer* repository.


docs/source/examples/config.txt
1.5, 8.0, 6, 5


CHANGES.rst
Changelog
=========

.. towncrier-draft-entries:: Not yet released

.. towncrier release notes start

0.3.9 (2022-03-04)
------------------

Features
^^^^^^^^

- Render GitHub Actions for continuous integration of the generated projects.
  (`#41 <https://github.com/csdms/babelizer/issues/41>`_)


Bugfixes
^^^^^^^^

- Fix minor formatting issues in rendered output. (`#49
  <https://github.com/csdms/babelizer/issues/49>`_)


Improved Documentation
^^^^^^^^^^^^^^^^^^^^^^

- Added paper for the Journal of Open Source Software (JOSS). (`#44
  <https://github.com/csdms/babelizer/issues/44>`_)
- Update docs for API changes. (`#47
  <https://github.com/csdms/babelizer/issues/47>`_)
- Update and clarify babelizer usage example. (`#54
  <https://github.com/csdms/babelizer/issues/54>`_)
- Complete API documentation. (`#55
  <https://github.com/csdms/babelizer/issues/55>`_)
- Improved the README documentation. (`#57
  <https://github.com/csdms/babelizer/issues/57>`_)
- Add coverage badge to the README file. (`#66
  <https://github.com/csdms/babelizer/issues/66>`_)


Misc
^^^^

- `#39 <https://github.com/csdms/babelizer/issues/39>`_, `#56
  <https://github.com/csdms/babelizer/issues/56>`_, `#65
  <https://github.com/csdms/babelizer/issues/65>`_, `#67
  <https://github.com/csdms/babelizer/issues/67>`_, `#68
  <https://github.com/csdms/babelizer/issues/68>`_


0.3.8 (2021-05-03)
------------------

Bugfixes
^^^^^^^^

- Fix minor formatting issues in rendered output. (`#49
  <https://github.com/csdms/babelizer/issues/49>`_)


Improved Documentation
^^^^^^^^^^^^^^^^^^^^^^

- Update docs for API changes. (`#47
  <https://github.com/csdms/babelizer/issues/47>`_)


0.3.7 (2021-03-19)
------------------

Features
^^^^^^^^

- Render GitHub Actions for continuous integration of the generated projects.
  (`#41 <https://github.com/csdms/babelizer/issues/41>`_)


Misc
^^^^

- Use GitHub Actions for continuous integration. (`#39
  <https://github.com/csdms/babelizer/issues/39>`_)


0.3.6 (2021-01-13)
------------------

Bugfixes
^^^^^^^^

- Fixed issue with metadata install path for Python components. (`#36
  <https://github.com/csdms/babelizer/issues/36>`_)


0.3.5 (2020-12-16)
------------------

Misc
^^^^

- Removed use of the toml package, instead use tomlkit. (`#34
  <https://github.com/csdms/babelizer/issues/34>`_)


0.3.4 (2020-12-16)
------------------

Bugfixes
^^^^^^^^

- Fixed a templating problem that caused Python BMIs to fail to build. (`#33
  <https://github.com/csdms/babelizer/issues/33>`_)


Improved Documentation
^^^^^^^^^^^^^^^^^^^^^^

- Minor edits to README and CLI help strings. (`#29
  <https://github.com/csdms/babelizer/issues/29>`_)
- Added text on the CSDMS Workbench to the README and docs. (`#30
  <https://github.com/csdms/babelizer/issues/30>`_)
- Clarified text in docs example. (`#31
  <https://github.com/csdms/babelizer/issues/31>`_)


Misc
^^^^

- Included the Python BMI example in the test suite.. (`#33
  <https://github.com/csdms/babelizer/issues/33>`_)


0.3.3 (2020-10-31)
------------------

Features
^^^^^^^^

- Added missing methods, primarily for unstructured grids, to C and C++
  implementation. (`#28 <https://github.com/csdms/babelizer/issues/28>`_)


Bugfixes
^^^^^^^^

- Fixed a rendering error that caused import lines to run together when
  wrapping multiple components (`#28
  <https://github.com/csdms/babelizer/issues/28>`_)


Misc
^^^^

- Removed Python 3.6 builds. (`#28 <https://github.com/csdms/babelizer/issues/28>`_)


0.3.2 (2020-10-08)
------------------

Bugfixes
^^^^^^^^

- Fix babelizing C++ libraries and added tests for C++ babelizing.  This fix
  necessitated a change to the "library" section of the babel.toml
  configuration file. (`#26
  <https://github.com/csdms/babelizer/issues/26>`_)


0.3.1 (2020-09-25)
------------------

Improved Documentation
^^^^^^^^^^^^^^^^^^^^^^

- Automatically set version in docs from the package version. (`#23 <https://github.com/csdms/babelizer/issues/23>`_)


0.3.0 (2020-09-24)
------------------

Improved Documentation
^^^^^^^^^^^^^^^^^^^^^^

- Added a user guide and API documentation (`#21 <https://github.com/csdms/babelizer/issues/21>`_)


0.2.0 (2020-09-10)
------------------

- Improved testing of the babelizer and added the bmi_heat C library to use for
  testing babelizing a package. (#20)

- Updated the babelized package for C libraries to BMI 2. (#20)

- Moved the external cookiecutter template into ``babelizer`` package.

- Renamed package to ``babelizer``.

- Added ``init``, ``update``, and ``quickstart`` subcommands to the babelize cli.

- Updated to use the new isort v5 api.

- Changed to use toml format by default for config files. The old yaml
  format is still understood but is deprecated.

- Fixed a bug where ``git init`` was called from the parent directory
  of the newly-created project, rather than from within the project.

- Removed versioneer from the babelized package. The version is now
  maintained within setup.py and releases should be made using
  zest.releaser.


0.1.3 (2018-10-28)
------------------

- Added new ``bmi-render`` command. The old ``bmi-wrap`` command is still available
  but will be removed in future releases.


0.1.2 (2018-06-28)
------------------

- Added update_until BMI method for C++.

- Fixed BMI return values for C.


0.1.1 (2018-04-09)
------------------

- Added continuous integration with Travis-CI. Build on Linux/MacOS with
  Python 2.7, 3.4, 3.5, 3.6.


0.1.0 (2018-04-05)
------------------

- Initial release


babelizer/cli.py
"""The command line interface to the babelizer."""

from __future__ import annotations

import fnmatch
import io
import os
import pathlib
import tempfile
from collections.abc import Collection
from functools import partial
from typing import cast

import click
import git

from babelizer._datadir import get_template_dir
from babelizer._files.gitignore import render as render_gitignore
from babelizer._files.license_rst import render as render_license
from babelizer._files.meson_build import render as render_meson_build
from babelizer._files.readme import render as render_readme
from babelizer._utils import get_setup_py_version
from babelizer._utils import save_files
from babelizer.config import BabelConfig
from babelizer.errors import OutputDirExistsError
from babelizer.errors import ScanError
from babelizer.errors import SetupPyError
from babelizer.errors import ValidationError
from babelizer.render import render

out = partial(click.secho, bold=True, err=True)
err = partial(click.secho, fg="red", err=True)


class BabelizerAbort(click.Abort):
    """Exception raised when a user interrupts the babelizer."""

    def __init__(self, message: str):
        err(message)


@click.group()
@click.version_option()
@click.option(
    "--cd",
    default=".",
    type=click.Path(exists=True, file_okay=False, dir_okay=True, readable=True),
    help="Change to directory, then execute.",
)
def babelize(cd: str) -> None:
    """Wrap BMI libraries with Python bindings."""
    os.chdir(cd)


@babelize.command()
@click.option(
    "-q",
    "--quiet",
    is_flag=True,
    help=(
        "Don't emit non-error messages to stderr. Errors are still emitted, "
        "silence those with 2>/dev/null"
    ),
)
@click.option(
    "-v", "--verbose", is_flag=True, help="Also emit status messages to stderr"
)
@click.option(
    "--template",
    default=None,
    help="Location of templates",
)
@click.option(
    "--package-version",
    default="0.1",
    help="The initial version of the babelized package",
)
@click.argument("meta", type=click.File(mode="r"))
def init(
    meta: click.File, template: str, quiet: bool, verbose: bool, package_version: str
) -> None:
    """Initialize a repository with babelized project files.

    META is babelizer configuration information, usually saved to a file.
    """
    template = template or get_template_dir()

    if not quiet:
        out(f"reading template from {template}")

    fmt = pathlib.Path(meta.name).suffix[1:] or "toml"
    try:
        babel_config = BabelConfig.from_stream(cast(io.TextIOBase, meta), fmt=fmt)
    except (ScanError, ValidationError) as error:
        raise BabelizerAbort(str(error))

    output = babel_config["package"]["name"]

    try:
        new_folder = render(
            babel_config,
            output,
            template=template,
            clobber=False,
            version=package_version,
        )
    except (ValidationError, OutputDirExistsError) as error:
        raise BabelizerAbort(str(error))

    if not quiet:
        out(
            "Don't forget to drop model metadata files into"
            f" {os.path.join(new_folder, 'meta')}"
        )

    repo = git.Repo(new_folder)
    repo.git.add("--all")
    repo.index.commit("Initial commit")

    print(new_folder)


@babelize.command()
@click.option(
    "-q",
    "--quiet",
    is_flag=True,
    help=(
        "Don't emit non-error messages to stderr. Errors are still emitted, "
        "silence those with 2>/dev/null."
    ),
)
@click.option(
    "-v", "--verbose", is_flag=True, help="Also emit status messages to stderr."
)
@click.option(
    "--template",
    default=None,
    help="Location of templates",
)
@click.option(
    "--set-version", default=None, help="Set the version of the updated package"
)
def update(
    template: str | None, quiet: bool, verbose: bool, set_version: str | None
) -> None:
    """Update an existing babelized project."""

    package_path = os.path.realpath(".")
    for fname in ("babel.toml", "babel.yaml", "plugin.yaml"):
        # if (package_path / fname).is_file():
        #     config_path = package_path / fname
        if os.path.isfile(os.path.join(package_path, fname)):
            config_path = os.path.join(package_path, fname)
            break
    else:
        config_path = None

    if not config_path:
        err("this does not appear to be a babelized folder (missing 'babel.toml')")
        raise click.Abort()

    template = template or get_template_dir()

    if not quiet:
        out(f"reading template from {template}")

    try:
        babel_config = BabelConfig.from_path(config_path)
    except ValidationError as error:
        raise BabelizerAbort(str(error))

    try:
        version = set_version or get_setup_py_version()
    except SetupPyError as error:
        raise BabelizerAbort(
            os.linesep.join(
                [
                    "the setup.py of this package has an error:",
                    f"{error}"
                    "unable to get package's version. Try using the '--set-version' option",
                ]
            )
        )
    version = "0.1.0" if version is None else version

    out(f"re-rendering {package_path}")
    with save_files(["CHANGES.rst", "CREDITS.rst"]):
        render(
            babel_config,
            os.path.dirname(package_path),
            # package_path.parent,
            template=template,
            clobber=True,
            version=version,
        )

    extra_files = _repo_contents(package_path) - _generated_files(
        babel_config, template=template, version=version
    )

    ignore = ["meta*", "notebooks*", "docs*", "**/data"]
    for pattern in ignore:
        extra_files.difference_update(fnmatch.filter(extra_files, pattern))

    if extra_files:
        out(f"found extra files in {package_path}:")
        for name in sorted(extra_files):
            out(f"  {name}")

    if not quiet:
        out(
            "Don't forget to drop model metadata files into"
            f" {os.path.join(package_path, 'meta')}"
        )

    print(package_path)


@babelize.command()
def sample_config() -> None:
    """Generate the babelizer configuration file."""
    print_sample_config()


@babelize.command()
def sample_license() -> None:
    """Generate a license file."""
    context = {
        "info": {
            "package_author": "Lyle Lanley",
            "summary": "A Monorail!",
            "package_license": "MIT License",
        }
    }
    print(render_license(context))


@babelize.command()
def sample_gitignore() -> None:
    """Generate a .gitignore file."""
    context = {
        "package": {"name": "springfield_monorail"},
        "library": {"monorail": {"language": "c"}},
    }
    print(render_gitignore(context))


@babelize.command()
@click.argument("extension", nargs=-1)
def sample_meson_build(extension: Collection[str]) -> None:
    """Generate a meson.build file."""
    if len(extension) == 0:
        contents = render_meson_build(
            [
                "springfield_monorail/lib/monorail.pyx",
                "springfield_monorail/lib/rail.pyx",
            ],
            install=[
                "springfield_monorail/__init__.py",
                "springfield_monorail/_bmi.py",
                "springfield_monorail/_version.py",
                "springfield_monorail/lib/__init__.py",
                "springfield_monorail/lib/monorail.pyx",
                "springfield_monorail/lib/rail.pyx",
            ],
        )
    else:
        contents = render_meson_build(extension)

    print(contents)


@babelize.command()
def sample_readme() -> None:
    context = {
        "language": "python",
        "open_source_license": "MIT License",
        "info": {
            "github_username": "lyle-lanley",
            "package_author": "Lyle Lanley",
            "summary": "A Monorail!",
            "package_license": "MIT License",
        },
        "components": {
            "Monorail": {"library": "monorail"},
            "Rail": {"library": "rail"},
        },
        "package": {"name": "springfield_monorail"},
    }
    print(render_readme(context))


def _get_dir_contents(base: str, trunk: str | None = None) -> set[str]:
    files = set()
    for item in os.listdir(base):
        if os.path.isdir(item):
            files |= _get_dir_contents(item, trunk=trunk)
        else:
            files.add(os.path.relpath(item, start=trunk))

    return files


def _repo_contents(base: str) -> set[str]:
    repo = git.Repo(base)
    return set(
        repo.git.ls_tree("--full-tree", "-r", "--name-only", "HEAD").splitlines()
    )


def _generated_files(
    babel_config: BabelConfig, template: str | None = None, version: str = "0.1"
) -> set[str]:
    with tempfile.TemporaryDirectory() as tmpdir:
        new_folder = render(
            babel_config,
            tmpdir,
            template=template,
            version=version,
        )
        return _get_dir_contents(new_folder, trunk=new_folder)


SAMPLE_CONFIG = """\
# See https://babelizer.readthedocs.io/ for more information

# Describe the library being wrapped.
[library.Monorail]
language = "c"
library = "bmimonorail"
header = "monorail.h"
entry_point = "register_monorail"

# Describe compiler options to build the library being wrapped.
[build]
undef_macros = []
define_macros = []
libraries = []
library_dirs = []
include_dirs = []
extra_compile_args = []

# Describe the newly wrapped package.
[package]
name = "springfield_monorail"
requirements = ["three_million_dollars"]

# Provide author and package information.
[info]
github_username = "lyle-lanley"
package_author = "Lyle Lanley"
package_author_email = "lyle@monorail.com"
package_license = "MIT License"
summary = '''
Well, sir, there's nothing on Earth like a genuine,
bona fide, electrified, six-car monorail. What'd I say?
Monorail! What's it called? Monorail! That's right! Monorail!
'''

# Set continuous integration options.
[ci]
python_version = [
    "3.10",
    "3.11",
    "3.12",
]
os = [
    "linux",
    "mac",
    "windows",
]
"""


def print_sample_config() -> int:
    """Print a sample babelizer configuration file."""
    print(SAMPLE_CONFIG, end="")
    return 0


if __name__ == "__main__":
    babelize(auto_envvar_prefix="BABELIZE")
